blog/_posts/2023-04-19-conda-r-geospatial.md

```diff
@@ -0,0 +1,87 @@
+To enter a Conda environment and have access to all the packages we've installed in it we use the command `conda activate ENV_NAME`, where `ENV_NAME` is the name of the environment. 
+However, on ARC4 we use the older version of this command `source activate ENV_NAME` to avoid some niggles we have with the newer behaviour of `conda activate`.
+
+```bash
+# activate the environment to use the packages you've installed
+source activate r-geo
+```
+
+After running `source activate` our prompt will be prepended with the name of our environment and we can now access the packages we installed into this environment.
+To install our R package we can either start up the R console and run `install.packages` directly or use the `R -e` command to pass in a string to the R to be run without opening a full R console.
+
+```bash
+# Install packages in R without starting an R console
+(r-geo) R -e 'install.packages("terra");' 
+```
+
+This will now attempt to install the package using the version of R we installed with Conda. 
+This ensures that all the libraries we installed through Conda will be available to R as it attempts to build any R packages that have C++ dependencies and uses the same compiler toolchain that was used to compile the version of R we installed with Conda.
+Any packages installed within a Conda environment are also saved within the Conda environment, helping isolate these R packages from any other R packages you may have installed using an R module or in another Conda environment.
+This again really helps with isolating our environment to prevent any package conflicts that occur and also makes it easy to dispose of our environment when we're done with it (we just delete the conda environment!).
+
+Overall, Conda is a very useful tool for managing dependencies and as demonstrated here is not just for Python!
+We can use it to manage versions of R, along with library dependencies for R packages. 
+This however does come with some potential drawbacks:
+- We've added another tool, by using Conda to manage dependencies we've added to the complexity of our project because for people to reproduce it they will now need Conda
+- We're constrained by what libraries are available via Conda, now that we're using Conda to manage our library dependencies we're dependent on someone making sure those libraries are available and up-to-date via Conda. It also becomes harder for us to compile a different library from source and make it available to our Conda-installed version of R
+- We're using a different compiler toolchain than the one on our system, on ARC4 the R module has been compiled with the compilers available on the HPC. When we use the Conda-installed version of R this has been compiled using a different compiler toolchain (that is also installed via Conda). This may not have optimised for the architecture of our HPC and so we might not be getting the best performance for Conda-installed R versus the R module on HPC
```

Comment:
Swings and roundabouts.  You may find the conda version has been more competently built, might use mkl or similar.  Would be interesting to do some benchmarks of real code.

Reply:
Maybe i'll add that as a homework!

Reply:
```suggestion
- We're using a different compiler toolchain than the one on our system, on ARC4 the R module has been compiled with the compilers available on the HPC. When we use the Conda-installed version of R this has been compiled using a different compiler toolchain (that is also installed via Conda). This may not have optimised for the architecture of our HPC and so we might not be getting the best performance for Conda-installed R versus the R module on HPC. Whilst out of scope of this blog post, you could have a go testing this by doing some benchmarking of code comparing Conda-installed R and the HPC R module.
```